homeassistant/components/gtfs/sensor.py

```diff
@@ -308,123 +382,303 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+                self._origin = stops[0]
+
+            if not self._destination:
+                stops = self._pygtfs.stops_by_id(self.destination)
+                if not stops:
+                    self._available = False
+                    _LOGGER.warning("Destination stop ID %s not found",
+                                    self.destination)
+                    return
+                self._destination = stops[0]
+
+            self._available = True
+
+            # Fetch next departure
             self._departure = get_next_departure(
                 self._pygtfs, self.origin, self.destination, self._offset,
                 self._include_tomorrow)
+
+            # Define the state as a UTC timestamp with ISO 8601 format
             if not self._departure:
                 self._state = None
-                self._attributes = {}
-                self._attributes['Info'] = "No more departures" if \
-                    self._include_tomorrow else "No more departures today"
-                if self._name == '':
-                    self._name = (self._custom_name or DEFAULT_NAME)
-                return
-
-            # Define the state as a UTC timestamp with ISO 8601 format.
-            arrival_time = dt_util.as_utc(
-                self._departure['arrival_time']).isoformat()
-            departure_time = dt_util.as_utc(
-                self._departure['departure_time']).isoformat()
-            self._state = departure_time
-
-            origin_station = self._departure['origin_station']
-            destination_station = self._departure['destination_station']
-            origin_stop_time = self._departure['origin_stop_time']
-            destination_stop_time = self._departure['destination_stop_time']
-            agency = self._departure['agency']
-            route = self._departure['route']
-            trip = self._departure['trip']
-
-            name = '{} {} to {} next departure'
+            else:
+                self._state = dt_util.as_utc(
+                    self._departure['departure_time']).isoformat()
+
+            # Fetch trip and route details once, unless updated
+            if not self._departure:
+                self._trip = None
+            else:
+                trip_id = self._departure['trip_id']
+                if not self._trip or self._trip.trip_id != trip_id:
+                    _LOGGER.info("Fetching trip details for %s", trip_id)
+                    self._trip = self._pygtfs.trips_by_id(trip_id)[0]
+
+                route_id = self._departure['route_id']
+                if not self._route or self._route.route_id != route_id:
+                    _LOGGER.info("Fetching route details for %s", route_id)
+                    self._route = self._pygtfs.routes_by_id(route_id)[0]
+
+            # Fetch agency details exactly once
+            if self._agency is None and self._route:
+                try:
+                    _LOGGER.info("Fetching agency details for %s",
```

Comment:
Same as above.

Reply:
Why wrap the logging in the try... except?

Reply:
Some agencies [do not set their agency_id in the `routes` table](https://transitfeeds.com/p/edmonton-transit-system/27/latest/file/routes.txt), so all values for the `agency_id` SQLite column end up as `None`. Prior to this PR [the update() would simply fail](https://github.com/home-assistant/home-assistant/blob/eb3355b05dd0edac38ee0e33a6d007b61c338fe9/homeassistant/components/sensor/gtfs.py#L169) if one didn't manually fix this missing relationship using something like:
```sql
UPDATE routes
SET agency_id = (
   SELECT agency_id
   FROM agency
   WHERE feed_id = 1
)
WHERE agency_id = 'None';
```

Now at least the sensor doesn't fail for such a non-critical metadata query, but a warning is issued about this missing relationship. A further PR could implement a pro-active fix at the database creation step.